Add composite template docs (#286)
operator-framework/olm-docs: content/en/docs/Reference/catalog-templates.md

Reviewer: L20-L21 need a new line added to list the new composite template.

Author: https://github.com/operator-framework/olm-docs/pull/286/commits/470a557691f687f8e566fcc0650697bf1e518631

### Usage
```sh
opm alpha render-template composite [flags]
```

| Flag                | Description                                                                            |
| ------------------- | -------------------------------------------------------------------------------------- |
| -f, --catalog-config string | File to use as the catalog configuration file (default "catalogs.yaml")        |
| -c, --composite-config string | File to use as the composite configuration file (default "catalog/config.yaml") |
| -h, --help          | help for composite                                                                     |
| -o, --output string | Output format (json|yaml) (default "json")                                             |
| --validate          | whether or not the created FBC should be validated (i.e 'opm validate') (default true) |
| --skip-tls-verify   | skip TLS certificate verification for container image registries while pulling bundles |
| --use-http          | use plain HTTP for container image registries while pulling bundles                    |

### Specifications

> The `composite template` is composed of two schemas that represent the operator author's role and the catalog maintainer's role.

Reviewer: This may venture into the nit-ty region, but I think it would be nice if we could introduce the sub-specs here and then refer to them throughout, instead of the role.  Maybe it's just me, but I get apostrophe fatigue. ;) 

```suggestion
The `composite template` is composed of two schemas that represent the operator author's role and the catalog maintainer's role, fulfilled by `olm.composite` and `olm.composite.catalogs`, respectively.
```

Author: https://github.com/operator-framework/olm-docs/pull/286/commits/8c89bd367d6835faed14561c8660ad7acb18e3fc

> #### Operator Author

Reviewer: Unfortunately, this h4 anchor is skipped by the current docsy config.  I was curious to see what would happen if we changed ./config.toml
```diff
diff --git a/config.toml b/config.toml
index 0b6a077..c026d18 100644
--- a/config.toml
+++ b/config.toml
@@ -34,6 +34,8 @@ pygmentsStyle = "tango"
   [markup.goldmark]
     [markup.goldmark.renderer]
       unsafe = true
+  [markup.tableOfContents]
+    endLevel = 4

 # Image processing configuration.
 [imaging]
 ```
 
 I have some concerns that it might yield too much noise in the right-hand TOC, but it would be nice to allow folks to dial in to this level, or maybe we want to find another way to structure this for similar effect.

Author: https://github.com/operator-framework/olm-docs/pull/286/commits/8c89bd367d6835faed14561c8660ad7acb18e3fc

> #### Operator Author

Author: https://github.com/operator-framework/olm-docs/pull/286/commits/8c89bd367d6835faed14561c8660ad7acb18e3fc

> Representing the operator author is the `olm.composite` schema which defines:

Reviewer 2: Maybe something like:
The `olm.composite` schema represents the operator author role and defines the following resources:

Author: Addressed in https://github.com/operator-framework/olm-docs/pull/286/commits/1cdb364fe5c9f65cbc47ff7effdb3c41bf104cc0

Reviewer 2: /resolve

- Where each input (`Component`) exists

> - How each contribution is mapped to a destination catalog (`Component.Name`). All destination catalogs defined in this file **must** also exist in the catalog configuration file created by the catalog maintainers

Reviewer: ```suggestion
- How each component is mapped to a destination catalog (`Component.Name`). All destination catalogs defined in this file **must** also exist in the catalog configuration file created by the catalog maintainers
```

Author: https://github.com/operator-framework/olm-docs/pull/286/commits/8c89bd367d6835faed14561c8660ad7acb18e3fc

- How each `Component` is processed to generate a catalog contribution (`Component.Strategy`)
- The directory structure the catalog contribution generation should follow (`Component.Destination`)

The cue schema for the `olm.composite` schema is:
```cue
#CompositeConfig: {
	Schema: "olm.composite"
	Components: [...#Component]
}

#Component: {
	Name:        string
	Destination: #ComponentDestination
	Strategy:  #BuildStrategy
}

#ComponentDestination {
	Path: string
}

#BuildStrategy: {
	Name:   string
	Template: #TemplateDefinition
}

#TemplateDefinition: {
	Schema: string
}
```

#### Catalog Maintainer

> Representing the catalog maintainer is the `olm.composite.catalogs` schema which defines:

Reviewer 2: Like above:
The `olm.composite.catalogs` schema represents the catalog maintainer role and defines the following resources:

Think resources is the wrong term in both cases. Perhaps, we can use _the following_ as a noun.

Author: Addressed in https://github.com/operator-framework/olm-docs/pull/286/commits/1cdb364fe5c9f65cbc47ff7effdb3c41bf104cc0

Reviewer 2: /resolve

- One or more catalogs (`Catalog`) identified by a string (`Catalog.Name`)
- Supported input formats (`Builder`) for each `Catalog` (`Catalog.Builders`) which is a list of strings where each item is a `Builder` schema. The currently supported `Builder` schemas are:
    - `olm.builder.basic` which represents that the use of [basic templates](#basic-template) for catalog contribution generation is allowed
    - `olm.builder.semver` which represents that the use of [semver templates](#semver-template) for catalog contribution generation is allowed

>     - `olm.builder.raw` which represents that the use of raw FBC for catalog contribution generation is allowed

Author: https://github.com/operator-framework/olm-docs/pull/286/commits/8c89bd367d6835faed14561c8660ad7acb18e3fc

    - `olm.builder.custom` which represents that the use of custom catalog contribution generation logic is allowed
- The expected directory structure of any generated contribution (`Catalog.Destination.WorkingDir`)

The cue schema for the `olm.composite.catalogs` schema is:
```cue
#CatalogConfig: {
	Schema: "olm.composite.catalogs"
	Catalogs: [... #Catalog]
}

#Catalog: {
	Name:        string
	Destination: #CatalogDestination
	Builders: [...string]
}

#CatalogDestination: {
	BaseImage:  string
	WorkingDir: string
}
```

### Examples
**catalogs.yaml**

> The following example specifies a catalog configuration that defines a catalog named `v1` and the contribution generation logic to only accepting input that uses the [basic template](#basic-template). When using the `opm alpha render-template composite` subcommand the path to this file can be specified with the `-f` option and can be a file path or a URL (the URL must return the raw file contents).

Reviewer 2: ```suggestion
The following example specifies a catalog configuration that defines a catalog named `v1`. This example limits the input to the [basic template](#basic-template) for the contribution generation logic. If you use the `opm alpha render-template composite` subcommand, the path to this file can be specified with the `-f` option. The option can be a file path or a URL. If you use a URL, the URL must return the raw file contents.
```
Just shortening some sentences and trying to clarify a couple of points where I got confused.

Author: Addressed in https://github.com/operator-framework/olm-docs/pull/286/commits/1cdb364fe5c9f65cbc47ff7effdb3c41bf104cc0

```yaml
schema: olm.composite.catalogs
catalogs:
- name: v1 {{< code_callout 1 >}}
  destination:
    baseImage: quay.io/operator-framework/opm:v1.24
    workingDir: catalogs/v1 {{< code_callout 2 >}}
  builders:

>     - olm.builder.basic {{< code_callout 3 >}}

Reviewer: I know it's a huge ask, but since elsewhere we're really trying to evangelize the semver template, could you remodel this to use semver instead of basic?

Author: https://github.com/operator-framework/olm-docs/pull/286/commits/8c89bd367d6835faed14561c8660ad7acb18e3fc

```

- {{< code_callout 1 >}} Defines the `v1` catalog configuration
- {{< code_callout 2 >}} Defines the directory path that all output should be placed for this catalog. In this example, when rendering using the composite catalog all output for the `v1` catalog will be put under the `catalogs/v1/` directory

> - {{< code_callout 3 >}} Defines the allowed builders for the `v1` catalog. In this example the only allowed builder is the basic builder, meaning the only contributions allowed to be rendered for this catalog are basic templates

Reviewer 2: ```suggestion
- {{< code_callout 3 >}} Defines the allowed builders for the `v1` catalog. In this example, the basic builder is the only allowed builder. As a result, this catalog renders allows only basic templates as contributions.
```
Trying to simplify, but maybe I lost the thread. Feel free to disregard if I am off base.

Author: Addressed in https://github.com/operator-framework/olm-docs/pull/286/commits/1cdb364fe5c9f65cbc47ff7effdb3c41bf104cc0

Reviewer 2: /resolve

**contributions.yaml**

The following example specifies a composite template configuration that defines the input, build process, and output of a catalog contribution for the `v1` catalog.

```yaml
schema: olm.composite
components:
- name: v1 {{< code_callout 1 >}}
  destination:
    path: my-package {{< code_callout 2 >}}
  strategy:
    name: basic
    template:
      schema: olm.builder.basic {{< code_callout 3 >}}
      config:
        input: components/v1.yaml {{< code_callout 4 >}}
        output: catalog.yaml {{< code_callout 5 >}}
```

- {{< code_callout 1 >}} Defines the build process for the `v1` catalog contribution.
- {{< code_callout 2 >}} Defines the name of the package being contributed and is used to create the package directory
- {{< code_callout 3 >}} Specifies that the input for the `v1` catalog contribution is a [basic template][#basic-template] so that the contribution can be rendered appropriately
- {{< code_callout 4 >}} Specifies the input file that should be used to generate the catalog contribution

> - {{< code_callout 5 >}} Defines what the name of the output file should be. This is joined as a path with the  `destination.path` value defined in {{< code_callout 2 >}}. In this example the full output path after rendering would be `my-package/catalog.yaml`

Reviewer 2: ```suggestion
- {{< code_callout 5 >}} Defines the name of the output file. This is joined as a path with the `destination.path` value defined in {{< code_callout 2 >}}. In this example, the full output path after rendering the template is `my-package/catalog.yaml`
```
Super small nits. The rationale behind avoiding _should_ is that it can imply both an imperative, "You should do this," or a probability, "This should happen." When it reads as a probability, the term can be misconstrued as hedging or a lack of certainty of the outcome.

Author: Addressed in https://github.com/operator-framework/olm-docs/pull/286/commits/1cdb364fe5c9f65cbc47ff7effdb3c41bf104cc0

Reviewer 2: /resolve

> When using the `opm alpha render-template composite -f catalogs.yaml -c contributions.yaml` command, the resulting output should look similar to:

Reviewer 2: In this instance, I think this is the right usage of _should_.

Here, a channel is generated for each template channel which differs by minor version, and each channel has a `replaces` edge from the predecessor channel to the next-lesser minor bundle version. Please note that at no time do we transgress across major-version boundaries with the channels, to be consistent with [the semver convention](https://semver.org/) for major versions, where the purpose is to make incompatible API changes.

## Composite Template

> A `composite template` is an input schema that accepts a configuration schema for generating a valid FBC for multiple catalogs at a time. This approach may be attractive to catalog maintainers looking to build multiple versions of a catalog while maintaining fewer files and operator authors looking to contribute their operator to multiple catalogs.

Reviewer 2: ```suggestion
You can use a `composite template` to generate a valid FBC for multiple catalogs at a time. A `composite template` is an input schema that accepts a configuration schema. Catalog maintainers can use a `composite template` to build multiple versions of a catalog while maintaining fewer files. Operator authors can use a `composite template` to contribute their operator to multiple catalogs.
```

Author: I like the sound of this much better, thanks!

### Usage
```sh
opm alpha render-template composite [flags] <filename>
```

| Flag                | Description                                                                            |
| ------------------- | -------------------------------------------------------------------------------------- |
| -h, --help          | prints help/usage information                                                          |
| -o, --output <type> | the output format, can be `yaml` or `json`                                             |
| -c, --composite-config <file> | File to use as the composite configuration file (default "catalog/config.yaml") |

> | --skip-tls-verify   | skip TLS certificate verification for container image registries while pulling bundles |

Reviewer 2: ```suggestion
| --skip-tls-verify   | skip TLS certificate verification for container image registries when pulling bundles |
```

Author: This flag description comes directly from the help text of the subcommand. You can see the help text here: https://github.com/operator-framework/operator-registry/blob/7835ea9c613ad1009cc36df8ad16e091d7775368/cmd/opm/root/cmd.go#L35

IMO we should keep this the same as the help text

Reviewer 2: /resolve

> | --use-http          | use plain HTTP for container image registries while pulling bundles                    |

Reviewer 2: ```suggestion
| --use-http          | use plain HTTP for container image registries when pulling bundles                    |
```

Reviewer 2: That makes sense. It is a minor style nit. :)

Reviewer 2: /resolve

| --validate          | whether or not the created FBC should be validated (i.e 'opm validate') (default true) |

> | --container-tool <tool> | container tool to be used when rendering veneers (should be an equivalent replacement to docker - similar to podman) (default "docker") |

Reviewer 2: ```suggestion
| --container-tool <tool> | container tool to be used when rendering veneers, such as Docker or Podman (default `docker`) |
```

Reviewer 2: /resolve

## Specification

> The composite veneer consists of a couple schemas:

Reviewer 2: ```suggestion
The composite veneer consists of the following schemas:
```

- `olm.composite`

>     - This schema is used to define the what, where, and how to build a particular contribution

Reviewer 2: ```suggestion
    - This schema defines what, where, and how to build a particular contribution
```

- `olm.composite.catalogs`

>     - This schema is used to define a catalog and the constraints for building the catalog

Reviewer 2: ```suggestion
    - This schema defines a catalog and the constraints for building the catalog
```

The `olm.composite.catalogs` cue schema is:
```cue
#CatalogConfig: {
	Schema: "olm.composite.catalogs"
	Catalogs: [... #Catalog]
}

#Catalog: {
	Name:        string
	Destination: #CatalogDestination
	Builders: [...string]
}

#CatalogDestination: {
	BaseImage:  string
	WorkingDir: string
}
```

The `olm.composite` cue schema is:
```
#CompositeConfig: {
	Schema: "olm.composite"
	Components: [...#Component]
}

#Component: {
	Name:        string
	Destination: #ComponentDestination
	Strategy:  #BuildStrategy
}

#ComponentDestination {
	Path: string
}

#BuildStrategy: {
	Name:   string
	Template: #TemplateDefinition
}

#TemplateDefinition: {
	Schema: string
	Config: _ @go(,json.RawMessage)
}
```

### Examples
**Catalog Config File**

> In this example, we are creating a catalog configuration that:

Reviewer 2: ```suggestion
The following example creates a catalog configuration that:
```

- Defines the `v4.10` catalog and restricts contributions to the catalog to only being `basic` templates
- Defines the `v4.11` catalog and restricts contributions to the catalog to only being `semver` templates
- Defines the `v4.12` catalog and restricts contributions to the catalog to only being raw FBC
- Defines the `v4.13` catalog and restricts contributions to the catalog to use a custom workflow

```yaml
schema: olm.composite.catalogs
catalogs:
- name: v4.10
  destination:
    baseImage: quay.io/operator-framework/opm:v1.24
    workingDir: catalogs/v4.10
  builders:
    - olm.builder.basic
- name: v4.11
  destination:
    baseImage: quay.io/operator-framework/opm:v1.25
    workingDir: catalogs/v4.11
  builders:
    - olm.builder.semver
- name: v4.12
  destination:
    baseImage: quay.io/operator-framework/opm:v1.26
    workingDir: catalogs/v4.12
  builders:
    - olm.builder.raw
- name: v4.13
  destination:
    baseImage: quay.io/operator-framework/opm:v1.26
    workingDir: catalogs/v4.13
  builders:
    - olm.builder.custom
```

**Composite Config File**

> In this example, we are creating a composite template configuration that:

Reviewer 2: ```suggestion
The following example creates a composite template configuration that:
```

> - Defines the input, build process, and output of a catalog contributions for `v4.10`, `v4.11`, `v4.12`, and `v4.13` catalogs

Reviewer 2: ```suggestion
- Defines the input, build process, and output of a catalog contributions for the `v4.10`, `v4.11`, `v4.12`, and `v4.13` catalogs
```

```yaml
schema: olm.composite
components:
- name: v4.10
  destination:
    path: my-package
  strategy:
    name: basic
    template:
      schema: olm.builder.basic
      config:
        input: components/v4.10.yaml 
        output: catalog.yaml
- name: v4.11
  destination:
    path: my-package
  strategy:
    name: semver
    template:
      schema: olm.builder.semver
      config:
        input: components/v4.11.yaml 
        output: catalog.yaml
- name: v4.12
  destination:
    path: my-package
  strategy:
    name: raw
    template:
      schema: olm.builder.raw
      config:
        input: components/v4.12.yaml 
        output: catalog.yaml
- name: v4.13
  destination:
    path: my-package
  strategy:
    name: custom
    template:
      schema: olm.builder.custom
      config:
        command: cat
        args:
          - "components/v4.13.yaml"
        output: catalog.yaml
```

> > **Note**: All catalogs defined in the composite template configuration file **must** also exist in the catalog configuration file. For example, if in the composite configuration we added a component for a catalog called `v4.20` the composite template rendering would fail as that catalog is not defined in the catalog configuration. 

Reviewer 2: ```suggestion
> **Note**: All catalogs defined in the composite template configuration file **must** also exist in the catalog configuration file. If the composite configuration includes a component that is not defined in the catalog configuration, the composite template will fail to render. 
```

Here, a channel is generated for each template channel which differs by minor version, and each channel has a `replaces` edge from the predecessor channel to the next-lesser minor bundle version. Please note that at no time do we transgress across major-version boundaries with the channels, to be consistent with [the semver convention](https://semver.org/) for major versions, where the purpose is to make incompatible API changes.

## Composite Template
You can use a `composite template` to generate a valid FBC for multiple catalogs at a time. A `composite template` is an input schema that accepts a configuration schema. Catalog maintainers can use a `composite template` to build multiple versions of a catalog while maintaining fewer files. Operator authors can use a `composite template` to contribute their operator to multiple catalogs.

### Usage
```sh
opm alpha render-template composite [flags] <filename>
```

| Flag                | Description                                                                            |
| ------------------- | -------------------------------------------------------------------------------------- |

> | -h, --help          | prints help/usage information                                                          |

Reviewer: somehow missing 
```  -f, --catalog-config string     File to use as the catalog configuration file (default "catalogs.yaml")```

| -o, --output <type> | the output format, can be `yaml` or `json`                                             |

> | -c, --composite-config <file> | File to use as the composite configuration file (default "catalog/config.yaml") |

Reviewer: @everettraven we need a PR to solve:
- changing (code, and now doc) references from 'catalog/config.yaml' to 'contributions.yaml'
- removing references to container tools (my bad!) (code, and now doc)

I can do it after the current queue of reviews, or you can do it and add to the queue. ;)

Author: Just to make sure I am following - is this a follow-up that needs to be done in both a new code change and docs change PR?

Reviewer: 2 independent PRs, since docs & code are discrete.  I was just making sure I captured the thesis so it didn't get lost.  Not scope for this PR, IMHO.

| --skip-tls-verify   | skip TLS certificate verification for container image registries while pulling bundles |
| --use-http          | use plain HTTP for container image registries while pulling bundles                    |
| --validate          | whether or not the created FBC should be validated (i.e 'opm validate') (default true) |
| --container-tool <tool> | container tool to be used when rendering veneers (should be an equivalent replacement to docker - similar to podman) (default "docker") |

## Specification
The composite veneer consists of the following schemas:

> - `olm.composite`

Reviewer: super-nit: can we ditch the sub-bullets for expansion details on the schema types?

>     - This schema defines what, where, and how to build a particular contribution

Reviewer: ```suggestion
    Representing the operator author, this schema defines where the contributions exist, how each contribution is mapped to a destination catalog (as specified by the catalog maintainer), and how each contribution is generated.
```

I like your brevity, but IMHO we need to be more descriptive of intent, and ideally do it in one place.  It's totally a preference thing, so I'm open for discussion on all points.

- `olm.composite.catalogs`

>     - This schema defines a catalog and the constraints for building the catalog

Reviewer: ```suggestion
    Representing the catalog maintainer, this schema defines one or more catalogs (`Catalog`), each with supported input formats (`Builders`) and the expected directory structure of any generated contribution (`WorkingDir`).
```

Reviewer: This is a likely location to give some details on the supported builders (and maybe their specs as well, though I confess I'm getting pretty stinking tired of CUE blobs).

> The `olm.composite.catalogs` cue schema is:

Reviewer: Any reason these aren't under their corresponding bullet, inline?
Any method to how they are in reverse order from above?

```cue
#CatalogConfig: {
	Schema: "olm.composite.catalogs"
	Catalogs: [... #Catalog]
}

#Catalog: {
	Name:        string
	Destination: #CatalogDestination
	Builders: [...string]
}

#CatalogDestination: {
	BaseImage:  string
	WorkingDir: string
}
```

The `olm.composite` cue schema is:

> ```

Reviewer: ```suggestion
```cue
```

#CompositeConfig: {
	Schema: "olm.composite"
	Components: [...#Component]
}

> #Component: {

Reviewer: We probably should include s/Component/Contribution/g in the "schema trim" jira. We weren't always consistent as we evolved this thing. ☹️

> #Component: {

Reviewer: I initially put in a comment to request we move from Component to Contribution nomenclature, but we need a term to refer to the proto-thing before it becomes a Contribution to the catalog ... and oh yeah, that's the Component.  
/me shakes head in disbelief. 

This means that we need to define Component in here, and consistently.  I attempted to retcon one of my earlier comments to include that info, and hopefully it isn't too bad.

	Name:        string
	Destination: #ComponentDestination
	Strategy:  #BuildStrategy
}

#ComponentDestination {
	Path: string
}

#BuildStrategy: {
	Name:   string
	Template: #TemplateDefinition
}

#TemplateDefinition: {
	Schema: string

> 	Config: _ @go(,json.RawMessage)

Reviewer: This is an implementation detail.  Is there a way for us to abstract this type?  Or is this the closest we can get in cue to an abstraction of the various logical types?  If we could get it as far as a minimum fit set similar to [_Meta](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#schema) that would be awesome.

Reviewer: Maybe we could capture verbiage here about the different schema to represent the different specs for the different BuildStrategy.Template.Schema types.

}
```

> ### Examples

Reviewer 3: I'm inclined to argue against advertising the raw and custom builders until we have an explicit ask from a user to add that.

> **Catalog Config File**

Reviewer 3: would be nice to point out that they can specify a remote location for that file (and hopefully we have one for the various kube versions you might be building catalogs for  - with the opm version that would be best suited for each)

> The following example creates a catalog configuration that:

Reviewer: ```suggestion
The following example specifies a catalog configuration that:
```

> - Defines the `v4.10` catalog and restricts contributions to the catalog to only being `basic` templates

Reviewer: I have mixed feelings about referring to OCP versions in an upstream doc.  Of course we understand the references, but there may be audiences who reach this doc who won't be able to infer the significance.  Would it be better if we abstracted like [v1, v2, v3] or [one, two, three]?

> - Defines the `v4.10` catalog and restricts contributions to the catalog to only being `basic` templates

Reviewer: ```suggestion
- Defines the `v4.10` catalog and restricts contributions to the catalog to only using [basic templates](#basic-template)
```

> - Defines the `v4.11` catalog and restricts contributions to the catalog to only being `semver` templates

Reviewer: ```suggestion
- Defines the `v4.11` catalog and restricts contributions to the catalog to only using [semver templates](#semver-template)
```

> - Defines the `v4.12` catalog and restricts contributions to the catalog to only being raw FBC

Reviewer: Internal to the template, we call this 'raw FBC', but to any audience outside the tool this is just 'FBC'.  Perhaps if we add a section in specifications which lists the supported builder types, this can just be a reference to that builder.

> - Defines the `v4.13` catalog and restricts contributions to the catalog to use a custom workflow

Reviewer: I think this line just underscores that we need to call out the different supported builder types with descriptions, because otherwise it's an exercise for the reader to figure out how to define an `olm.builder.custom` object. 😞

```yaml
schema: olm.composite.catalogs
catalogs:

> - name: v4.10 {{< code_callout 1 >}}

Reviewer 3: nit: OCP reference in upstream docs

  destination:
    baseImage: quay.io/operator-framework/opm:v1.24
    workingDir: catalogs/v4.10 {{< code_callout 2 >}}
  builders:
    - olm.builder.basic {{< code_callout 3 >}}
- name: v4.11
  destination:
    baseImage: quay.io/operator-framework/opm:v1.25
    workingDir: catalogs/v4.11
  builders:

>     - olm.builder.semver {{< code_callout 4 >}}

Reviewer 3: would be good to show that this can be a list of builders.
```suggestion
    - olm.builder.basic
    - olm.builder.semver {{< code_callout 4 >}}
```

- name: v4.12
  destination:
    baseImage: quay.io/operator-framework/opm:v1.26
    workingDir: catalogs/v4.12
  builders:
    - olm.builder.raw {{< code_callout 5 >}}
- name: v4.13
  destination:
    baseImage: quay.io/operator-framework/opm:v1.26
    workingDir: catalogs/v4.13
  builders:
    - olm.builder.custom {{< code_callout 6 >}}
```

- {{< code_callout 1 >}} Defines the `v4.10` catalog configuration
- {{< code_callout 2 >}} Defines the directory path that all output should be placed for this catalog. In this example, when rendering using the composite catalog all output for the `v4.10` catalog will be put under the `catalogs/v4.10/` directory

> - {{< code_callout 3 >}} Defines the allowed builders for the `v4.10` catalog. In this example the only allowed builder is the basic builder, meaning the only contributions allowed to be rendered for this catalog are basic templates

Reviewer: >  ...meaning the only contributions allowed to be rendered for this catalog are basic templates

IMHO, these requirements/restrictions should be detailed in the schema definitions so folks can look in one place for all the interactions

- {{< code_callout 4 >}} Definition for the semver builder, which is used to render semver templates

> - {{< code_callout 5 >}} Definition for the raw builder, which is used to render full FBC (no templates)

Reviewer: Is it correct to say 'render full FBC' since rendering is how we produce FBC?  Not life-threatening, but maybe "handle raw FBC"?

> - {{< code_callout 6 >}} Definition for the custom builder, which is used to render FBC using custom methods

Reviewer: Def more argument for details on the supported builders, so folks can construct them.

**Composite Config File**

The following example creates a composite template configuration that:

> - Defines the input, build process, and output of a catalog contributions for the `v4.10`, `v4.11`, `v4.12`, and `v4.13` catalogs

Reviewer: Why is this a bulleted list?

- {{< code_callout 7 >}} Specifies the contribution for the `v4.12` catalog is already full FBC
- {{< code_callout 8 >}} Specifies the contribution for the `v4.13` catalog requires running custom commands that result in a full FBC being printed to STDOUT. In this example we are just running the `cat` command on a file that is full FBC

Using the above examples, the `components/` directory should look similar to:
```tree
components
├── v4.10.yaml
├── v4.11.yaml
├── v4.12.yaml
└── v4.13.yaml
```

For more details about the contents of each file, expand the corresponding dropdown:
<details>
<summary>components/v4.10.yaml</summary>

```yaml
---
defaultChannel: preview
name: webhook-operator
schema: olm.package
---
image: quay.io/olmtest/webhook-operator-bundle:0.0.3
name: webhook-operator.v0.0.1
schema: olm.bundle
---
schema: olm.channel
package: webhook-operator
name: preview
entries:
  - name: webhook-operator.v0.0.1
```

</details>

<details>
<summary>components/v4.11.yaml</summary>

```yaml
Schema: olm.semver
GenerateMajorChannels: true
GenerateMinorChannels: true
Stable:
  Bundles:
  - Image: quay.io/olmtest/webhook-operator-bundle:0.0.3
```

</details>

<details>

> <summary>components/v4.12.yaml</summary>

Reviewer: I'm not convinced that we need to go to the level of detail in the example of providing all the content of all the participants.  On the other hand, I can see us donating the example repo to OF and making a reference at the top of the example that we're referring to LOCATION for the example. 

If we do find all this detail useful, I think it's sufficient if we strip out all `olm.bundle.object` objects and make a note that we're filtering them out (similar to examples for other template types on this page).
